sql/schemachanger/scplan: TestPlanAlterTable failed

sql/schemachanger/scplan.TestPlanAlterTable [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3098883&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3098883&tab=artifacts#/) on master @ [69308cce3bb7e660908cb3e2724eedd271ce5585](https://github.com/cockroachdb/cockroach/commits/69308cce3bb7e660908cb3e2724eedd271ce5585):


```
             TableID: 63
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 54
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 57
         Stage 1
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 54
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 57
           *scop.CreateGcJobForDescriptor
             DescID: 55
           *scop.CreateGcJobForDescriptor
             DescID: 58
           *scop.CreateGcJobForDescriptor
        @@ -84,14 +80,14 @@
             DescID: 56
           *scop.CreateGcJobForDescriptor
             DescID: 62
           *scop.CreateGcJobForDescriptor
             DescID: 63
        -  *scop.DrainDescriptorName
        -    TableID: 54
        -  *scop.DrainDescriptorName
        -    TableID: 57
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 54
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 57
           *scop.DrainDescriptorName
             TableID: 55
           *scop.DrainDescriptorName
             TableID: 58
           *scop.DrainDescriptorName
        @@ -108,10 +104,14 @@
             TableID: 62
           *scop.DrainDescriptorName
             TableID: 63
           *scop.MarkDescriptorAsDropped
             TableID: 53
        +  *scop.DrainDescriptorName
        +    TableID: 54
        +  *scop.DrainDescriptorName
        +    TableID: 57
           *scop.MarkDescriptorAsDropped
             TableID: 52
         Stage 2
           *scop.CreateGcJobForDescriptor
             DescID: 53
    --- FAIL: TestPlanAlterTable/drop_database (0.47s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestPlanAlterTable PKG=./pkg/sql/schemachanger/scplan TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</p>
</details>
/cc @cockroachdb/sql-queries fqazi
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPlanAlterTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/schemachanger/scplan.TestPlanAlterTable [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3098884&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3098884&tab=artifacts#/) on master @ [69308cce3bb7e660908cb3e2724eedd271ce5585](https://github.com/cockroachdb/cockroach/commits/69308cce3bb7e660908cb3e2724eedd271ce5585):


```
             TableID: 63
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 54
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 57
         Stage 1
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 54
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 57
           *scop.CreateGcJobForDescriptor
             DescID: 55
           *scop.CreateGcJobForDescriptor
             DescID: 58
           *scop.CreateGcJobForDescriptor
        @@ -84,14 +80,14 @@
             DescID: 56
           *scop.CreateGcJobForDescriptor
             DescID: 62
           *scop.CreateGcJobForDescriptor
             DescID: 63
        -  *scop.DrainDescriptorName
        -    TableID: 54
        -  *scop.DrainDescriptorName
        -    TableID: 57
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 54
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 57
           *scop.DrainDescriptorName
             TableID: 55
           *scop.DrainDescriptorName
             TableID: 58
           *scop.DrainDescriptorName
        @@ -108,10 +104,14 @@
             TableID: 62
           *scop.DrainDescriptorName
             TableID: 63
           *scop.MarkDescriptorAsDropped
             TableID: 53
        +  *scop.DrainDescriptorName
        +    TableID: 54
        +  *scop.DrainDescriptorName
        +    TableID: 57
           *scop.MarkDescriptorAsDropped
             TableID: 52
         Stage 2
           *scop.CreateGcJobForDescriptor
             DescID: 53
    --- FAIL: TestPlanAlterTable/drop_database (2.63s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestPlanAlterTable PKG=./pkg/sql/schemachanger/scplan TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</p>
</details>
/cc @cockroachdb/sql-queries fqazi
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPlanAlterTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/schemachanger/scplan.TestPlanAlterTable [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3099729&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3099729&tab=artifacts#/) on master @ [0e747e4ac3630a9b41f2eccb7ace0e2d17258ae6](https://github.com/cockroachdb/cockroach/commits/0e747e4ac3630a9b41f2eccb7ace0e2d17258ae6):


```
             TableID: 63
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 54
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 57
         Stage 1
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 54
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 57
           *scop.CreateGcJobForDescriptor
             DescID: 55
           *scop.CreateGcJobForDescriptor
             DescID: 58
           *scop.CreateGcJobForDescriptor
        @@ -84,14 +80,14 @@
             DescID: 56
           *scop.CreateGcJobForDescriptor
             DescID: 62
           *scop.CreateGcJobForDescriptor
             DescID: 63
        -  *scop.DrainDescriptorName
        -    TableID: 54
        -  *scop.DrainDescriptorName
        -    TableID: 57
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 54
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 57
           *scop.DrainDescriptorName
             TableID: 55
           *scop.DrainDescriptorName
             TableID: 58
           *scop.DrainDescriptorName
        @@ -108,10 +104,14 @@
             TableID: 62
           *scop.DrainDescriptorName
             TableID: 63
           *scop.MarkDescriptorAsDropped
             TableID: 53
        +  *scop.DrainDescriptorName
        +    TableID: 54
        +  *scop.DrainDescriptorName
        +    TableID: 57
           *scop.MarkDescriptorAsDropped
             TableID: 52
         Stage 2
           *scop.CreateGcJobForDescriptor
             DescID: 53
    --- FAIL: TestPlanAlterTable/drop_database (1.44s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestPlanAlterTable PKG=./pkg/sql/schemachanger/scplan TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/sql-queries fqazi
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPlanAlterTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/schemachanger/scplan.TestPlanAlterTable [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3100549&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3100549&tab=artifacts#/) on master @ [0e747e4ac3630a9b41f2eccb7ace0e2d17258ae6](https://github.com/cockroachdb/cockroach/commits/0e747e4ac3630a9b41f2eccb7ace0e2d17258ae6):


```
             TableID: 63
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 54
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 57
         Stage 1
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 54
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 57
           *scop.CreateGcJobForDescriptor
             DescID: 55
           *scop.CreateGcJobForDescriptor
             DescID: 58
           *scop.CreateGcJobForDescriptor
        @@ -84,14 +80,14 @@
             DescID: 56
           *scop.CreateGcJobForDescriptor
             DescID: 62
           *scop.CreateGcJobForDescriptor
             DescID: 63
        -  *scop.DrainDescriptorName
        -    TableID: 54
        -  *scop.DrainDescriptorName
        -    TableID: 57
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 54
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 57
           *scop.DrainDescriptorName
             TableID: 55
           *scop.DrainDescriptorName
             TableID: 58
           *scop.DrainDescriptorName
        @@ -108,10 +104,14 @@
             TableID: 62
           *scop.DrainDescriptorName
             TableID: 63
           *scop.MarkDescriptorAsDropped
             TableID: 53
        +  *scop.DrainDescriptorName
        +    TableID: 54
        +  *scop.DrainDescriptorName
        +    TableID: 57
           *scop.MarkDescriptorAsDropped
             TableID: 52
         Stage 2
           *scop.CreateGcJobForDescriptor
             DescID: 53
    --- FAIL: TestPlanAlterTable/drop_database (0.53s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestPlanAlterTable PKG=./pkg/sql/schemachanger/scplan TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</p>
</details>
/cc @cockroachdb/sql-queries fqazi
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPlanAlterTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/schemachanger/scplan.TestPlanAlterTable [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3100550&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3100550&tab=artifacts#/) on master @ [0e747e4ac3630a9b41f2eccb7ace0e2d17258ae6](https://github.com/cockroachdb/cockroach/commits/0e747e4ac3630a9b41f2eccb7ace0e2d17258ae6):


```
             TableID: 63
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 54
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 57
         Stage 1
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 54
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 57
           *scop.CreateGcJobForDescriptor
             DescID: 55
           *scop.CreateGcJobForDescriptor
             DescID: 58
           *scop.CreateGcJobForDescriptor
        @@ -84,14 +80,14 @@
             DescID: 56
           *scop.CreateGcJobForDescriptor
             DescID: 62
           *scop.CreateGcJobForDescriptor
             DescID: 63
        -  *scop.DrainDescriptorName
        -    TableID: 54
        -  *scop.DrainDescriptorName
        -    TableID: 57
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 54
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 57
           *scop.DrainDescriptorName
             TableID: 55
           *scop.DrainDescriptorName
             TableID: 58
           *scop.DrainDescriptorName
        @@ -108,10 +104,14 @@
             TableID: 62
           *scop.DrainDescriptorName
             TableID: 63
           *scop.MarkDescriptorAsDropped
             TableID: 53
        +  *scop.DrainDescriptorName
        +    TableID: 54
        +  *scop.DrainDescriptorName
        +    TableID: 57
           *scop.MarkDescriptorAsDropped
             TableID: 52
         Stage 2
           *scop.CreateGcJobForDescriptor
             DescID: 53
    --- FAIL: TestPlanAlterTable/drop_database (2.81s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestPlanAlterTable PKG=./pkg/sql/schemachanger/scplan TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</p>
</details>
/cc @cockroachdb/sql-queries fqazi
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPlanAlterTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/schemachanger/scplan.TestPlanAlterTable [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3101162&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3101162&tab=artifacts#/) on master @ [fe729f712066a8ab1ec6c12ed8a2becc2a14b910](https://github.com/cockroachdb/cockroach/commits/fe729f712066a8ab1ec6c12ed8a2becc2a14b910):


```
             TableID: 63
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 54
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 57
         Stage 1
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 54
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 57
           *scop.CreateGcJobForDescriptor
             DescID: 55
           *scop.CreateGcJobForDescriptor
             DescID: 58
           *scop.CreateGcJobForDescriptor
        @@ -84,14 +80,14 @@
             DescID: 56
           *scop.CreateGcJobForDescriptor
             DescID: 62
           *scop.CreateGcJobForDescriptor
             DescID: 63
        -  *scop.DrainDescriptorName
        -    TableID: 54
        -  *scop.DrainDescriptorName
        -    TableID: 57
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 54
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 57
           *scop.DrainDescriptorName
             TableID: 55
           *scop.DrainDescriptorName
             TableID: 58
           *scop.DrainDescriptorName
        @@ -108,10 +104,14 @@
             TableID: 62
           *scop.DrainDescriptorName
             TableID: 63
           *scop.MarkDescriptorAsDropped
             TableID: 53
        +  *scop.DrainDescriptorName
        +    TableID: 54
        +  *scop.DrainDescriptorName
        +    TableID: 57
           *scop.MarkDescriptorAsDropped
             TableID: 52
         Stage 2
           *scop.CreateGcJobForDescriptor
             DescID: 53
    --- FAIL: TestPlanAlterTable/drop_database (1.27s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestPlanAlterTable PKG=./pkg/sql/schemachanger/scplan TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/sql-queries fqazi
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPlanAlterTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/schemachanger/scplan.TestPlanAlterTable [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3101541&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3101541&tab=artifacts#/) on master @ [6e6fda134ea844007c59ba0b3fe3758be9737833](https://github.com/cockroachdb/cockroach/commits/6e6fda134ea844007c59ba0b3fe3758be9737833):


```
             TableID: 63
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 54
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 57
         Stage 1
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 54
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 57
           *scop.CreateGcJobForDescriptor
             DescID: 55
           *scop.CreateGcJobForDescriptor
             DescID: 58
           *scop.CreateGcJobForDescriptor
        @@ -84,14 +80,14 @@
             DescID: 56
           *scop.CreateGcJobForDescriptor
             DescID: 62
           *scop.CreateGcJobForDescriptor
             DescID: 63
        -  *scop.DrainDescriptorName
        -    TableID: 54
        -  *scop.DrainDescriptorName
        -    TableID: 57
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 54
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 57
           *scop.DrainDescriptorName
             TableID: 55
           *scop.DrainDescriptorName
             TableID: 58
           *scop.DrainDescriptorName
        @@ -108,10 +104,14 @@
             TableID: 62
           *scop.DrainDescriptorName
             TableID: 63
           *scop.MarkDescriptorAsDropped
             TableID: 53
        +  *scop.DrainDescriptorName
        +    TableID: 54
        +  *scop.DrainDescriptorName
        +    TableID: 57
           *scop.MarkDescriptorAsDropped
             TableID: 52
         Stage 2
           *scop.CreateGcJobForDescriptor
             DescID: 53
    --- FAIL: TestPlanAlterTable/drop_database (3.34s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestPlanAlterTable PKG=./pkg/sql/schemachanger/scplan TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/sql-queries fqazi
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPlanAlterTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/schemachanger/scplan.TestPlanAlterTable [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3105141&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3105141&tab=artifacts#/) on master @ [1da7cca4105195edd954b5cf0a1e3154c36afcb3](https://github.com/cockroachdb/cockroach/commits/1da7cca4105195edd954b5cf0a1e3154c36afcb3):


```
             TableID: 63
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 54
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 57
         Stage 1
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 54
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 57
           *scop.CreateGcJobForDescriptor
             DescID: 55
           *scop.CreateGcJobForDescriptor
             DescID: 58
           *scop.CreateGcJobForDescriptor
        @@ -84,14 +80,14 @@
             DescID: 56
           *scop.CreateGcJobForDescriptor
             DescID: 62
           *scop.CreateGcJobForDescriptor
             DescID: 63
        -  *scop.DrainDescriptorName
        -    TableID: 54
        -  *scop.DrainDescriptorName
        -    TableID: 57
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 54
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 57
           *scop.DrainDescriptorName
             TableID: 55
           *scop.DrainDescriptorName
             TableID: 58
           *scop.DrainDescriptorName
        @@ -108,10 +104,14 @@
             TableID: 62
           *scop.DrainDescriptorName
             TableID: 63
           *scop.MarkDescriptorAsDropped
             TableID: 53
        +  *scop.DrainDescriptorName
        +    TableID: 54
        +  *scop.DrainDescriptorName
        +    TableID: 57
           *scop.MarkDescriptorAsDropped
             TableID: 52
         Stage 2
           *scop.CreateGcJobForDescriptor
             DescID: 53
    --- FAIL: TestPlanAlterTable/drop_database (0.99s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestPlanAlterTable PKG=./pkg/sql/schemachanger/scplan TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/sql-queries fqazi
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPlanAlterTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/schemachanger/scplan.TestPlanAlterTable [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3104699&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3104699&tab=artifacts#/) on master @ [1da7cca4105195edd954b5cf0a1e3154c36afcb3](https://github.com/cockroachdb/cockroach/commits/1da7cca4105195edd954b5cf0a1e3154c36afcb3):


```
             TableID: 63
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 54
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 57
         Stage 1
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 54
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 57
           *scop.CreateGcJobForDescriptor
             DescID: 55
           *scop.CreateGcJobForDescriptor
             DescID: 58
           *scop.CreateGcJobForDescriptor
        @@ -84,14 +80,14 @@
             DescID: 56
           *scop.CreateGcJobForDescriptor
             DescID: 62
           *scop.CreateGcJobForDescriptor
             DescID: 63
        -  *scop.DrainDescriptorName
        -    TableID: 54
        -  *scop.DrainDescriptorName
        -    TableID: 57
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 54
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 57
           *scop.DrainDescriptorName
             TableID: 55
           *scop.DrainDescriptorName
             TableID: 58
           *scop.DrainDescriptorName
        @@ -108,10 +104,14 @@
             TableID: 62
           *scop.DrainDescriptorName
             TableID: 63
           *scop.MarkDescriptorAsDropped
             TableID: 53
        +  *scop.DrainDescriptorName
        +    TableID: 54
        +  *scop.DrainDescriptorName
        +    TableID: 57
           *scop.MarkDescriptorAsDropped
             TableID: 52
         Stage 2
           *scop.CreateGcJobForDescriptor
             DescID: 53
    --- FAIL: TestPlanAlterTable/drop_database (0.53s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestPlanAlterTable PKG=./pkg/sql/schemachanger/scplan TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</p>
</details>
/cc @cockroachdb/sql-queries fqazi
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPlanAlterTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/schemachanger/scplan.TestPlanAlterTable [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3104700&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3104700&tab=artifacts#/) on master @ [1da7cca4105195edd954b5cf0a1e3154c36afcb3](https://github.com/cockroachdb/cockroach/commits/1da7cca4105195edd954b5cf0a1e3154c36afcb3):


```
             TableID: 63
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 54
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 57
         Stage 1
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 54
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 57
           *scop.CreateGcJobForDescriptor
             DescID: 55
           *scop.CreateGcJobForDescriptor
             DescID: 58
           *scop.CreateGcJobForDescriptor
        @@ -84,14 +80,14 @@
             DescID: 56
           *scop.CreateGcJobForDescriptor
             DescID: 62
           *scop.CreateGcJobForDescriptor
             DescID: 63
        -  *scop.DrainDescriptorName
        -    TableID: 54
        -  *scop.DrainDescriptorName
        -    TableID: 57
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 54
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 57
           *scop.DrainDescriptorName
             TableID: 55
           *scop.DrainDescriptorName
             TableID: 58
           *scop.DrainDescriptorName
        @@ -108,10 +104,14 @@
             TableID: 62
           *scop.DrainDescriptorName
             TableID: 63
           *scop.MarkDescriptorAsDropped
             TableID: 53
        +  *scop.DrainDescriptorName
        +    TableID: 54
        +  *scop.DrainDescriptorName
        +    TableID: 57
           *scop.MarkDescriptorAsDropped
             TableID: 52
         Stage 2
           *scop.CreateGcJobForDescriptor
             DescID: 53
    --- FAIL: TestPlanAlterTable/drop_database (2.59s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestPlanAlterTable PKG=./pkg/sql/schemachanger/scplan TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</p>
</details>
/cc @cockroachdb/sql-queries fqazi
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPlanAlterTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/schemachanger/scplan.TestPlanAlterTable [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3107009&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3107009&tab=artifacts#/) on master @ [1bd0fe95cc0ee93faeb1711bdd118c708e1e2e1a](https://github.com/cockroachdb/cockroach/commits/1bd0fe95cc0ee93faeb1711bdd118c708e1e2e1a):


```
             TableID: 63
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 54
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 57
         Stage 1
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 54
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 57
           *scop.CreateGcJobForDescriptor
             DescID: 55
           *scop.CreateGcJobForDescriptor
             DescID: 58
           *scop.CreateGcJobForDescriptor
        @@ -84,14 +80,14 @@
             DescID: 56
           *scop.CreateGcJobForDescriptor
             DescID: 62
           *scop.CreateGcJobForDescriptor
             DescID: 63
        -  *scop.DrainDescriptorName
        -    TableID: 54
        -  *scop.DrainDescriptorName
        -    TableID: 57
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 54
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 57
           *scop.DrainDescriptorName
             TableID: 55
           *scop.DrainDescriptorName
             TableID: 58
           *scop.DrainDescriptorName
        @@ -108,10 +104,14 @@
             TableID: 62
           *scop.DrainDescriptorName
             TableID: 63
           *scop.MarkDescriptorAsDropped
             TableID: 53
        +  *scop.DrainDescriptorName
        +    TableID: 54
        +  *scop.DrainDescriptorName
        +    TableID: 57
           *scop.MarkDescriptorAsDropped
             TableID: 52
         Stage 2
           *scop.CreateGcJobForDescriptor
             DescID: 53
    --- FAIL: TestPlanAlterTable/drop_database (1.70s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestPlanAlterTable PKG=./pkg/sql/schemachanger/scplan TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/sql-queries fqazi
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPlanAlterTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/schemachanger/scplan.TestPlanAlterTable [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3107213&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3107213&tab=artifacts#/) on master @ [1bd0fe95cc0ee93faeb1711bdd118c708e1e2e1a](https://github.com/cockroachdb/cockroach/commits/1bd0fe95cc0ee93faeb1711bdd118c708e1e2e1a):


```
             TableID: 63
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 54
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 57
         Stage 1
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 54
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 57
           *scop.CreateGcJobForDescriptor
             DescID: 55
           *scop.CreateGcJobForDescriptor
             DescID: 58
           *scop.CreateGcJobForDescriptor
        @@ -84,14 +80,14 @@
             DescID: 56
           *scop.CreateGcJobForDescriptor
             DescID: 62
           *scop.CreateGcJobForDescriptor
             DescID: 63
        -  *scop.DrainDescriptorName
        -    TableID: 54
        -  *scop.DrainDescriptorName
        -    TableID: 57
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 54
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 57
           *scop.DrainDescriptorName
             TableID: 55
           *scop.DrainDescriptorName
             TableID: 58
           *scop.DrainDescriptorName
        @@ -108,10 +104,14 @@
             TableID: 62
           *scop.DrainDescriptorName
             TableID: 63
           *scop.MarkDescriptorAsDropped
             TableID: 53
        +  *scop.DrainDescriptorName
        +    TableID: 54
        +  *scop.DrainDescriptorName
        +    TableID: 57
           *scop.MarkDescriptorAsDropped
             TableID: 52
         Stage 2
           *scop.CreateGcJobForDescriptor
             DescID: 53
    --- FAIL: TestPlanAlterTable/drop_database (1.34s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestPlanAlterTable PKG=./pkg/sql/schemachanger/scplan TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/sql-queries fqazi
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPlanAlterTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/schemachanger/scplan.TestPlanAlterTable [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3107530&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3107530&tab=artifacts#/) on master @ [1bd0fe95cc0ee93faeb1711bdd118c708e1e2e1a](https://github.com/cockroachdb/cockroach/commits/1bd0fe95cc0ee93faeb1711bdd118c708e1e2e1a):


```
        +    TableID: 54
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 57
         Stage 1
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 54
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 57
           *scop.CreateGcJobForDescriptor
             DescID: 55
           *scop.CreateGcJobForDescriptor
             DescID: 58
           *scop.CreateGcJobForDescriptor
        @@ -84,14 +80,14 @@
             DescID: 56
           *scop.CreateGcJobForDescriptor
             DescID: 62
           *scop.CreateGcJobForDescriptor
             DescID: 63
        -  *scop.DrainDescriptorName
        -    TableID: 54
        -  *scop.DrainDescriptorName
        -    TableID: 57
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 54
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 57
           *scop.DrainDescriptorName
             TableID: 55
           *scop.DrainDescriptorName
             TableID: 58
           *scop.DrainDescriptorName
        @@ -108,10 +104,14 @@
             TableID: 62
           *scop.DrainDescriptorName
             TableID: 63
           *scop.MarkDescriptorAsDropped
             TableID: 53
        +  *scop.DrainDescriptorName
        +    TableID: 54
        +  *scop.DrainDescriptorName
        +    TableID: 57
           *scop.MarkDescriptorAsDropped
             TableID: 52
         Stage 2
           *scop.CreateGcJobForDescriptor
             DescID: 53
E210622 22:02:49.187195 4258 kv/kvserver/consistency_queue.go:191  [n1,consistencyChecker,s1,r20/1:/Table/2{4-5}] 1  computing own checksum: could not dial node ID 1: failed to connect to n1 at 127.0.0.1:38355: stopped
E210622 22:02:49.187460 4258 kv/kvserver/queue.go:1093  [n1,consistencyChecker,s1,r20/1:/Table/2{4-5}] 2  computing own checksum: could not dial node ID 1: failed to connect to n1 at 127.0.0.1:38355: stopped
    --- FAIL: TestPlanAlterTable/drop_database (1.11s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestPlanAlterTable PKG=./pkg/sql/schemachanger/scplan TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/sql-queries fqazi
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPlanAlterTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/schemachanger/scplan.TestPlanAlterTable [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3108165&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3108165&tab=artifacts#/) on master @ [ebaee33e27ffa79814359ed4bd059d99c45c74a0](https://github.com/cockroachdb/cockroach/commits/ebaee33e27ffa79814359ed4bd059d99c45c74a0):


```
             TableID: 63
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 54
        +  *scop.MarkDescriptorAsDropped
        +    TableID: 57
         Stage 1
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 54
        -  *scop.CreateGcJobForDescriptor
        -    DescID: 57
           *scop.CreateGcJobForDescriptor
             DescID: 55
           *scop.CreateGcJobForDescriptor
             DescID: 58
           *scop.CreateGcJobForDescriptor
        @@ -84,14 +80,14 @@
             DescID: 56
           *scop.CreateGcJobForDescriptor
             DescID: 62
           *scop.CreateGcJobForDescriptor
             DescID: 63
        -  *scop.DrainDescriptorName
        -    TableID: 54
        -  *scop.DrainDescriptorName
        -    TableID: 57
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 54
        +  *scop.CreateGcJobForDescriptor
        +    DescID: 57
           *scop.DrainDescriptorName
             TableID: 55
           *scop.DrainDescriptorName
             TableID: 58
           *scop.DrainDescriptorName
        @@ -108,10 +104,14 @@
             TableID: 62
           *scop.DrainDescriptorName
             TableID: 63
           *scop.MarkDescriptorAsDropped
             TableID: 53
        +  *scop.DrainDescriptorName
        +    TableID: 54
        +  *scop.DrainDescriptorName
        +    TableID: 57
           *scop.MarkDescriptorAsDropped
             TableID: 52
         Stage 2
           *scop.CreateGcJobForDescriptor
             DescID: 53
    --- FAIL: TestPlanAlterTable/drop_database (1.24s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestPlanAlterTable PKG=./pkg/sql/schemachanger/scplan TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/sql-queries fqazi
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPlanAlterTable.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/logictest/testdata/logic_test/new_schema_changer
+++ b/pkg/sql/logictest/testdata/logic_test/new_schema_changer
@@ -641,3 +641,8 @@ DROP DATABASE db1
 
 statement ok
 DROP DATABASE db1 CASCADE
+
+# Works around view back references to tables not being cleaned up.
+# see issue #66677
+statement ok
+SET experimental_use_new_schema_changer = 'off'

--- a/pkg/sql/schemachanger/scbuild/database.go
+++ b/pkg/sql/schemachanger/scbuild/database.go
@@ -12,8 +12,10 @@ package scbuild
 
 import (
 	"context"
+	"sort"
 
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/privilege"
@@ -40,9 +42,17 @@ func (b *buildContext) dropDatabase(ctx context.Context, n *tree.DropDatabase) {
 	if err != nil {
 		panic(err)
 	}
-
-	dropIDs := catalog.DescriptorIDSet{}
+	// Order schemas to guarantee the order in which operations
+	// are generated for testing.
+	schemaIDs := make([]descpb.ID, 0, len(schemas))
 	for schemaID := range schemas {
+		schemaIDs = append(schemaIDs, schemaID)
+	}
+	sort.SliceStable(schemaIDs, func(i, j int) bool {
+		return schemaIDs[i] < schemaIDs[j]
+	})
+	dropIDs := catalog.DescriptorIDSet{}
+	for _, schemaID := range schemaIDs {
 		dropIDs.Add(schemaID)
 		schemaDesc, err := b.Descs.GetImmutableSchemaByID(ctx, b.EvalCtx.Txn, schemaID, tree.SchemaLookupFlags{Required: true})
 		if err != nil {

--- a/pkg/sql/schemachanger/scbuild/schema.go
+++ b/pkg/sql/schemachanger/scbuild/schema.go
@@ -12,6 +12,7 @@ package scbuild
 
 import (
 	"context"
+	"sort"
 
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/dbdesc"
@@ -51,6 +52,11 @@ func (b *buildContext) dropSchemaDesc(
 		panic(pgerror.Newf(pgcode.DependentObjectsStillExist,
 			"schema %q is not empty and CASCADE was not specified", sc.GetName()))
 	}
+	// Sort the dropped IDs to ensure a consistent order of
+	// operations.
+	sort.SliceStable(dropIDs, func(i, j int) bool {
+		return dropIDs[i] < dropIDs[j]
+	})
 	schemaNode := scpb.Schema{
 		SchemaID:         sc.GetID(),
 		DependentObjects: dropIDs,

--- a/pkg/sql/schemachanger/scbuild/testdata/drop_database
+++ b/pkg/sql/schemachanger/scbuild/testdata/drop_database
@@ -107,14 +107,14 @@ DROP DATABASE db1 CASCADE
   state: PUBLIC
   details:
     dependentObjects:
-    - 63
     - 55
     - 56
-    - 62
     - 58
     - 59
     - 60
     - 61
+    - 62
+    - 63
     - 64
     schemaId: 53
 - DROP Sequence:{DescID: 54}

--- a/pkg/sql/schemachanger/scbuild/testdata/drop_schema
+++ b/pkg/sql/schemachanger/scbuild/testdata/drop_schema
@@ -71,14 +71,14 @@ DROP SCHEMA defaultdb.SC1 CASCADE
   state: PUBLIC
   details:
     dependentObjects:
-    - 61
     - 54
     - 55
-    - 60
     - 56
     - 57
     - 58
     - 59
+    - 60
+    - 61
     - 62
     schemaId: 52
 - DROP Sequence:{DescID: 54}

--- a/pkg/sql/schemachanger/scplan/rules.go
+++ b/pkg/sql/schemachanger/scplan/rules.go
@@ -266,9 +266,6 @@ var rules = map[scpb.Element]targetRules{
 					nextState: scpb.State_ABSENT,
 					op: func(this *scpb.Database) []scop.Op {
 						ops := []scop.Op{
-							&scop.CreateGcJobForDescriptor{
-								DescID: this.DatabaseID,
-							},
 							&scop.DrainDescriptorName{TableID: this.DatabaseID},
 						}
 						return ops
@@ -334,9 +331,6 @@ var rules = map[scpb.Element]targetRules{
 					nextState: scpb.State_ABSENT,
 					op: func(this *scpb.Schema) []scop.Op {
 						ops := []scop.Op{
-							&scop.CreateGcJobForDescriptor{
-								DescID: this.SchemaID,
-							},
 							&scop.DrainDescriptorName{TableID: this.SchemaID},
 						}
 						return ops
@@ -511,9 +505,6 @@ var rules = map[scpb.Element]targetRules{
 					nextState: scpb.State_ABSENT,
 					op: func(this *scpb.Type) []scop.Op {
 						ops := []scop.Op{
-							&scop.CreateGcJobForDescriptor{
-								DescID: this.TypeID,
-							},
 							&scop.DrainDescriptorName{TableID: this.TypeID},
 						}
 						return ops

--- a/pkg/sql/schemachanger/scplan/testdata/drop_database
+++ b/pkg/sql/schemachanger/scplan/testdata/drop_database
@@ -134,10 +134,6 @@ Stage 1
     DescID: 64
   *scop.CreateGcJobForDescriptor
     DescID: 56
-  *scop.CreateGcJobForDescriptor
-    DescID: 62
-  *scop.CreateGcJobForDescriptor
-    DescID: 63
   *scop.DrainDescriptorName
     TableID: 54
   *scop.DrainDescriptorName
@@ -165,10 +161,6 @@ Stage 1
   *scop.MarkDescriptorAsDropped
     TableID: 52
 Stage 2
-  *scop.CreateGcJobForDescriptor
-    DescID: 53
-  *scop.CreateGcJobForDescriptor
-    DescID: 52
   *scop.DrainDescriptorName
     TableID: 53
   *scop.DrainDescriptorName

--- a/pkg/sql/schemachanger/scplan/testdata/drop_schema
+++ b/pkg/sql/schemachanger/scplan/testdata/drop_schema
@@ -147,10 +147,6 @@ Stage 1
     DescID: 61
   *scop.CreateGcJobForDescriptor
     DescID: 54
-  *scop.CreateGcJobForDescriptor
-    DescID: 59
-  *scop.CreateGcJobForDescriptor
-    DescID: 60
   *scop.DrainDescriptorName
     TableID: 53
   *scop.DrainDescriptorName
@@ -172,7 +168,5 @@ Stage 1
   *scop.MarkDescriptorAsDropped
     TableID: 52
 Stage 2
-  *scop.CreateGcJobForDescriptor
-    DescID: 52
   *scop.DrainDescriptorName
     TableID: 52

--- a/pkg/sql/schemachanger/scplan/testdata/drop_type
+++ b/pkg/sql/schemachanger/scplan/testdata/drop_type
@@ -11,10 +11,6 @@ Stage 0
   *scop.MarkDescriptorAsDropped
     TableID: 53
 Stage 1
-  *scop.CreateGcJobForDescriptor
-    DescID: 52
-  *scop.CreateGcJobForDescriptor
-    DescID: 53
   *scop.DrainDescriptorName
     TableID: 52
   *scop.DrainDescriptorName

